docs(pt-br): Translate you-might-not-need-an-effect.md (#675)
reactjs/pt-br.react.dev: src/content/learn/you-might-not-need-an-effect.md

---
title: 'Talvez você não precise de um Effect'
---

<Intro>

> Effects são um escape do paradigma do React. Eles te permitem "dar um passo para fora" do React e sincronizar seus componentes com algum serviço externo, como um widget não React, a rede ou o DOM do navegador. Se não houver nenhum sistema externo envolvido (por exemplo, se você quiser atualizar o estado de um componente quando algumas props ou estado mudarem), você não deveria precisar de um Effect. Remover Effects desnecessários tornará seu código mais compreensível, mais rápido de executar e menos propenso a erros.

Reviewer: ```suggestion
Effects são um escape do paradigma do React. Eles te permitem "dar um passo para fora" do React e sincronizar seus componentes com algum serviço externo, como um widget não React, a rede ou o DOM do navegador. Se não houver nenhum sistema externo envolvido (por exemplo, se você quiser atualizar o estado de um componente quando algumas props ou state mudarem), você não deveria precisar de um Effect. Remover Effects desnecessários tornará seu código mais compreensível, mais rápido de executar e menos propenso a erros.
```

}
```

> Isso é mais complicado que o necessário. É ineficente também: isso faz uma renderização inteira com um valor desatualizado de `fullName`, e depois imediatamente re-renderiza com o valor atualizado. Remova a variável de estado e o Effect:

Author: "state variable" traduziria para "variável de state", "variável state" ou só "state"?

> Isso é mais complicado que o necessário. É ineficente também: isso faz uma renderização inteira com um valor desatualizado de `fullName`, e depois imediatamente re-renderiza com o valor atualizado. Remova a variável de estado e o Effect:

Author: ```suggestion
Isso é mais complicado que o necessário. É ineficente também: isso faz uma renderização inteira com um valor desatualizado de `fullName`, e depois imediatamente re-renderiza com o valor atualizado. Remova o state e o Effect:
```

```js
console.time('filter array');
const visibleTodos = useMemo(() => {
  return getFilteredTodos(todos, filter); // Ignorado se `todos` e `filter` não tiverem sido alterados
}, [todos, filter]);
console.timeEnd('filter array');
```

`useMemo` não fará com que a *primeira* renderização seja mais rápida. Ele só ajuda a evitar trabalho desnecessário nas atualizações.

Lembre-se de que seu computador provavelmente é mais rápido do que o dos seus usuários, portanto, é uma boa ideia testar o desempenho com uma lentidão artificial. Por exemplo, o Chrome oferece uma opção de [Throttling de CPU](https://developer.chrome.com/blog/new-in-devtools-61/#throttling) para isso.

Observe também que medir o desempenho no desenvolvimento não fornecerá os resultados mais precisos. (Por exemplo, quando o [Strict Mode](/reference/react/StrictMode)  estiver ativado, você verá cada componente ser renderizado duas vezes em vez de uma). Para obter os tempos mais precisos, faça uma build de produção de seu app e teste-o em um dispositivo parecido com o de seus usuários.

</DeepDive>

### Redefinir todo o estados quando uma prop é modificada {/*resetting-all-state-when-a-prop-changes*/}

> Esse componente `ProfilePage` recebe uma propriedade `userId`. A página contém um input de comentário e você usa uma variável de estado `comment` para manter seu valor. Um dia, você percebeu um problema: quando você navega de um perfil para outro, o estado `comment` não é redefinido. Como resultado, é fácil publicar acidentalmente um comentário no perfil de um usuário errado. Para corrigir o problema, você deseja limpar a variável de estado `comment` sempre que o `userId` for alterado:

Author: Aqui pode ser um campo? É input no sentido de caixa de texto mesmo

}

function Profile({ userId }) {
  // ✅ Este e qualquer outro estado abaixo serão redefinidos automaticamente na mudança de chave
  const [comment, setComment] = useState('');
  // ...
}
```

Normalmente, o React preserva o estado quando o mesmo componente é renderizado no mesmo local. **Ao passar `userId` como `key` para o componente `Profile`, você está pedindo ao React para tratar dois componentes `Profile` com `userId` diferentes como dois componentes diferentes que não devem compartilhar nenhum estado.** Sempre que a chave (que você definiu como `userId`) mudar, o React irá recriar o DOM e [redefinir o estado](/learn/preserving-and-resetting-state#option-2-resetting-state-with-a-key) do componente `Profile` e todos os seus filhos. Agora o campo `comment` será apagado automaticamente ao navegar entre os perfis.

Observe que, neste exemplo, somente o componente externo `ProfilePage` é exportado e visível para outros arquivos do projeto. Os componentes que renderizam o `ProfilePage` não precisam passar a chave para ele: eles passam o `userId` como uma propriedade normal. O fato de `ProfilePage` passar a chave para o componente `Profile` interno é um detalhe de implementação.

> ### Ajustando algum estado quando uma prop é alterada {/*adjusting-some-state-when-a-prop-changes*/}

Reviewer: ```suggestion
### Ajustando algum state quando uma prop é alterada {/*adjusting-some-state-when-a-prop-changes*/}
```

> Às vezes, você pode querer redefinir ou ajustar uma parte do estado em uma alteração de prop, mas não todo ele.

Reviewer: ```suggestion
Às vezes, você pode querer redefinir ou ajustar uma parte do state em uma alteração de prop, mas não todo ele.
```

Author: ```suggestion
Às vezes, você pode querer redefinir ou ajustar algum state específico, sem afetar outros, quando uma prop for alterada.
```

> Esse componente `List` recebe uma lista de `items` como uma prop e mantém o item selecionado na variável de estado `selection`. Você deseja redefinir a `selection` para `null` sempre que a prop `items` receber um array diferente:

Reviewer: ```suggestion
Este componente `List` recebe uma lista de `items` como uma prop e mantém o item selecionado na variável de estado `selection`. Você deseja redefinir a `selection` para `null` sempre que a prop `items` receber um array diferente:
```

```js {5-8}
function List({ items }) {
  const [isReverse, setIsReverse] = useState(false);
  const [selection, setSelection] = useState(null);

  // 🔴 Evite: Ajustar o estado na mudança de prop em um Effect
  useEffect(() => {
    setSelection(null);
  }, [items]);
  // ...
}
```

> Isso também não é ideal. Toda vez que os `items` mudarem, a `List` e seus componentes filhos serão renderizados com um valor `selection` desatualizado no início. Em seguida, o React atualizará o DOM e executará os Effects. Por fim, a chamada `setSelection(null)` causará outra re-renderização da `List` e de seus componentes filhos, reiniciando todo o processo novamente.

Reviewer: ```suggestion
Isto também não é ideal. Toda vez que os `items` mudarem, a `List` e seus componentes filhos serão renderizados com um valor `selection` desatualizado no início. Em seguida, o React atualizará o DOM e executará os Effects. Por fim, a chamada `setSelection(null)` causará outra re-renderização da `List` e de seus componentes filhos, reiniciando todo o processo novamente.
```

> Comece excluindo o Effect. Em vez disso, ajuste o estado diretamente durante a renderização:

Reviewer: ```suggestion
Comece excluindo o Effect. Em vez disso, ajuste o state diretamente durante a renderização:
```

```js {5-11}
function List({ items }) {
  const [isReverse, setIsReverse] = useState(false);
  const [selection, setSelection] = useState(null);

>   // Melhor: Ajustar o estado durante a renderização

Reviewer: ```suggestion
  // Melhor: Ajustar o state durante a renderização
```

}
```

> [Armazenar informações de renderizações anteriores](/reference/react/useState#storing-information-from-previous-renders) dessa maneira pode ser difícil de entender, mas é melhor do que atualizar o mesmo estado em um Effect. No exemplo acima, `setSelection` é chamado diretamente durante uma renderização. O React irá re-renderizar a `List` *imediatamente* após sair com uma instrução `return`. O React ainda não renderizou os filhos da `List` ou atualizou o DOM, então isso permite que os filhos da `List` pulem a renderização do valor obsoleto da `selection`.

Reviewer: ```suggestion
[Armazenar informações de renderizações anteriores](/reference/react/useState#storing-information-from-previous-renders) dessa maneira pode ser difícil de entender, mas é melhor do que atualizar o mesmo state em um Effect. No exemplo acima, `setSelection` é chamado diretamente durante uma renderização. O React irá re-renderizar a `List` *imediatamente* após sair com uma instrução `return`. O React ainda não renderizou os filhos da `List` ou atualizou o DOM, então isso permite que os filhos da `List` pulem a renderização do valor obsoleto da `selection`.
```

> Quando você atualiza um componente durante a renderização, o React joga fora o JSX retornado e imediatamente reinicia a renderização. Para evitar repetições em cascata muito lentas, o React só permite que você atualize o estado do *mesmo* componente durante uma renderização. Se você atualizar o estado de outro componente durante uma renderização, verá um erro. Uma condição como `items !== prevItems` é necessária para evitar loops. Você pode ajustar o estado dessa forma, mas quaisquer outros efeitos colaterais (como alterar o DOM ou definir timeouts) devem ficar em event handlers ou Effects para [manter os componentes puros](/learn/keeping-components-pure).

Reviewer: ```suggestion
Quando você atualiza um componente durante a renderização, o React descarta o JSX retornado e imediatamente reinicia a renderização. Para evitar repetições em cascata muito lentas, o React só permite que você atualize o state do *mesmo* componente durante uma renderização. Se você atualizar o state de outro componente durante uma renderização, verá um erro. Uma condição como `items !== prevItems` é necessária para evitar loops. Você pode ajustar o state dessa forma, mas quaisquer outros efeitos colaterais (como alterar o DOM ou definir timeouts) devem ficar em manipuladores de evento ou Effects para [manter os componentes puros](/learn/keeping-components-pure).
```

> **Embora esse padrão seja mais eficiente do que um Effect, a maioria dos componentes também não deve precisar dele.** Não importa como você o faça, o ajuste do estado com base em props ou outro estado torna o fluxo de dados mais difícil de entender e depurar. Sempre verifique se, em vez disso, você pode [redefinir todo o estado com uma chave](#resetting-all-state-when-a-prop-changes) ou [calcular tudo durante a renderização](#updating-state-based-on-props-or-state). Por exemplo, em vez de armazenar (e redefinir) o *item* selecionado, você pode armazenar o *ID do item selecionado:*

Reviewer: ```suggestion
**Embora esse padrão seja mais eficiente do que um Effect, a maioria dos componentes também não deve precisar dele.** Não importa como você o faça, o ajuste do state com base em props ou outro state torna o fluxo de dados mais difícil de entender e depurar. Sempre verifique se, em vez disso, você pode [redefinir todo o estado com uma chave](#resetting-all-state-when-a-prop-changes) ou [calcular tudo durante a renderização](#updating-state-based-on-props-or-state). Por exemplo, em vez de armazenar (e redefinir) o *item* selecionado, você pode armazenar o *ID do item selecionado:*
```

```js {3-5}
function List({ items }) {
  const [isReverse, setIsReverse] = useState(false);
  const [selectedId, setSelectedId] = useState(null);
  // ✅ Melhor: calcular tudo durante a renderização
  const selection = items.find(item => item.id === selectedId) ?? null;
  // ...
}
```

> Agora não há necessidade de "ajustar" o estado. Se o item com o ID selecionado estiver na lista, ele permanecerá selecionado. Se não estiver, a `selection` calculada durante a renderização será `null` porque nenhum item correspondente foi encontrado. Esse comportamento é diferente, mas indiscutivelmente melhor porque a maioria das alterações nos `items` preserva a seleção.

Reviewer: ```suggestion
Agora não há necessidade de "ajustar" o state. Se o item com o ID selecionado estiver na lista, ele permanecerá selecionado. Se não estiver, a `selection` calculada durante a renderização será `null` porque nenhum item correspondente foi encontrado. Esse comportamento é diferente, mas indiscutivelmente melhor porque a maioria das alterações nos `items` preserva a seleção.
```

> ### Compartilhamento de lógica entre event handlers {/*sharing-logic-between-event-handlers*/}

Reviewer: ```suggestion
### Compartilhamento de lógica entre manipuladores de evento {/*sharing-logic-between-event-handlers*/}
```

> Digamos que você tenha uma página de produto com dois botões (Buy e Checkout) que permitem que você compre o produto. Você deseja exibir uma notificação sempre que o usuário colocar o produto no carrinho. Chamar `showNotification()` nos handlers de cliques dos dois botões parece repetitivo, portanto, você pode se sentir tentado a colocar essa lógica em um Effect:

Reviewer: ```suggestion
Digamos que você tenha uma página de produto com dois botões (Buy e Checkout) que permitem que você compre o produto. Você deseja exibir uma notificação sempre que o usuário colocar o produto no carrinho. Chamar `showNotification()` nos manipuladores de cliques dos dois botões parece repetitivo, portanto, você pode se sentir tentado a colocar essa lógica em um Effect:
```

}
```

> Esse Effect é desnecessário. Ele também provavelmente causará bugs. Por exemplo, digamos que seu aplicativo "lembre" o carrinho de compras entre os recarregamentos da página. Se você adicionar um produto ao carrinho uma vez e atualizar a página, a notificação aparecerá novamente. Ela continuará aparecendo toda vez que você atualizar a página desse produto. Isso ocorre porque `product.isInCart` já será `true` no carregamento da página, de modo que o Effect acima chamará `showNotification()`.

Author: ```suggestion
Esse Effect é desnecessário. Ele também provavelmente causará bugs. Por exemplo, digamos que sua aplicação "lembre" o carrinho de compras entre os recarregamentos da página. Se você adicionar um produto ao carrinho uma vez e atualizar a página, a notificação aparecerá novamente. Ela continuará aparecendo toda vez que você atualizar a página desse produto. Isso ocorre porque `product.isInCart` já será `true` no carregamento da página, de modo que o Effect acima chamará `showNotification()`.
```

> **Quando não tiver certeza se algum código deve estar em um Effect ou em um manipulador de eventos, pergunte a si mesmo *por que* esse código precisa ser executado. Use Effects somente para códigos que devem ser executados *porque* o componente foi exibido ao usuário.** Neste exemplo, a notificação deve aparecer porque o usuário *pressionou o botão*, não porque a página foi exibida! Exclua o Effect e coloque a lógica compartilhada em uma função chamada de ambos os event handlers:

Reviewer: ```suggestion
**Quando não tiver certeza se algum código deve estar em um Effect ou em um manipulador de eventos, pergunte a si mesmo *por que* esse código precisa ser executado. Use Effects somente para códigos que devem ser executados *porque* o componente foi exibido ao usuário.** Neste exemplo, a notificação deve aparecer porque o usuário *pressionou o botão*, não porque a página foi exibida! Exclua o Effect e coloque a lógica compartilhada em uma função chamada de ambos os manipuladores de evento:
```

```js {2-6,9,13}
function ProductPage({ product, addToCart }) {

>   // ✅ Bom: A lógica específica do evento é chamada pelos event handlers

Reviewer: ```suggestion
  // ✅ Bom: A lógica específica do evento é chamada pelos manipuladores
```

}
```

Isso remove o Effect desnecessário e corrige o bug.

### Enviando uma solicitação POST {/*sending-a-post-request*/}

> Esse componente `Form` envia dois tipos de solicitações POST. Ele envia um evento de analytics quando é montado. Quando você preencher o formulário e clicar no botão Submit, ele enviará uma solicitação POST para o endpoint `/api/register`:

Reviewer: ```suggestion
Este componente `Form` envia dois tipos de solicitações POST. Ele envia um evento de analytics quando é montado. Quando você preencher o formulário e clicar no botão Submit, ele enviará uma solicitação POST ao endpoint `/api/register`:
```

}
```

Vamos aplicar os mesmos critérios do exemplo anterior.

A solicitação POST de análise deve permanecer em um Effect. Isso ocorre porque o _motivo_ para enviar o evento de análise é que o formulário foi exibido. (Ele seria disparado duas vezes no desenvolvimento, mas [veja aqui](/learn/synchronizing-with-effects#enviar-análises) para saber como lidar com isso).

> No entanto, a solicitação POST `/api/register` não é causada pelo formulário que está sendo _exibido_. Você só deseja enviar a solicitação em um momento específico: quando o usuário pressiona o botão. Isso só deve acontecer _naquela interação específica_. Exclua o segundo Effect e mova essa solicitação POST para o event handler:

Reviewer: ```suggestion
No entanto, a solicitação POST `/api/register` não é causada pelo formulário que está sendo _exibido_. Você só deseja enviar a solicitação em um momento específico: quando o usuário pressiona o botão. Isso só deve acontecer _naquela interação específica_. Exclua o segundo Effect e mova essa solicitação POST para o manipulador de evento:
```

```js {12-13}
function Form() {
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');

  // ✅ Bom: Essa lógica é executada porque o componente foi exibido
  useEffect(() => {
    post('/analytics/event', { eventName: 'visit_form' });
  }, []);

  function handleSubmit(e) {
    e.preventDefault();

>     // ✅ Bom: A lógica específica do evento está no event handler.

Reviewer: ```suggestion
    // ✅ Bom: A lógica específica do evento está no manipulador.
```

    post('/api/register', { firstName, lastName });
  }
  // ...
}
```

> Ao decidir se deve colocar alguma lógica em um event handler ou em um Effect, a principal pergunta que precisa ser respondida é _que tipo de lógica_ ela é da perspectiva do usuário. Se essa lógica for causada por uma interação específica, mantenha-a no event handler. Se for causada pelo fato de o usuário _ver_ o componente na tela, mantenha-a no Effect.

Reviewer: ```suggestion
Ao decidir se deve colocar alguma lógica em um manipulador de evento ou em um Effect, a principal pergunta que precisa ser respondida é _que tipo de lógica_ ela é da perspectiva do usuário. Se essa lógica for causada por uma interação específica, mantenha-a no manipulador de evento. Se for causada pelo fato de o usuário _ver_ o componente na tela, mantenha-a no Effect.
```

### Cadeias de processamentos {/*chains-of-computations*/}

> Às vezes, você pode se sentir tentado a encadear Effects que ajustam uma parte do estado com base em outro estado:

Reviewer: ```suggestion
Às vezes, você pode se sentir tentado a encadear Effects que ajustam uma parte do state com base em outro state:
```

Author: ```suggestion
Às vezes, você pode se sentir tentado a encadear Effects que ajustam um state com base em outro state:
```

  const [round, setRound] = useState(1);
  const [isGameOver, setIsGameOver] = useState(false);

>   // 🔴 Evitar: Cadeias de Effects que ajustam o estado apenas para acionar uns aos outros

Reviewer: ```suggestion
  // 🔴 Evitar: Cadeias de Effects que ajustam o state apenas para acionar uns aos outros
```

  // ...
```

Há dois problemas com esse código.

Um problema é que ele é muito ineficiente: o componente (e seus filhos) precisa ser renderizado novamente entre cada chamada `set` na cadeia. No exemplo acima, na pior das hipóteses (`setCard` → render → `setGoldCardCount` → render → `setRound` → render → `setIsGameOver` → render), há três re-renderizações desnecessárias da árvore abaixo.

> Mesmo que isso não fosse lento, à medida que seu código evolui, você se depara com casos em que a "cadeia" que você escreveu não atende aos novos requisitos. Imagine que você esteja adicionando uma maneira de percorrer o histórico dos movimentos do jogo. Você faria isso atualizando cada variável de estado para um valor do passado. Entretanto, definir o estado `card` como um valor do passado acionaria a cadeia de Effects novamente e alteraria os dados que você está mostrando. Esse tipo de código costuma ser rígido e frágil.

Reviewer: ```suggestion
Mesmo que isso não fosse lento, à medida que seu código evolui, você se depara com casos em que a "cadeia" que você escreveu não atende aos novos requisitos. Imagine que você esteja adicionando uma maneira de percorrer o histórico dos movimentos do jogo. Você faria isso atualizando cada variável de state para um valor do passado. Entretanto, definir o state `card` como um valor do passado acionaria a cadeia de Effects novamente e alteraria os dados que você está mostrando. Esse tipo de código costuma ser rígido e frágil.
```

> Nesse caso, é melhor calcular o que for possível durante a renderização e ajustar o estado no event handler:

Reviewer: ```suggestion
Neste caso, é melhor calcular o que for possível durante a renderização e ajustar o state no manipulador de evento:
```

```js {6-7,14-26}
function Game() {
  const [card, setCard] = useState(null);
  const [goldCardCount, setGoldCardCount] = useState(0);
  const [round, setRound] = useState(1);

  // ✅ Calcular o que puder durante a renderização
  const isGameOver = round > 5;

  function handlePlaceCard(nextCard) {
    if (isGameOver) {
      throw Error('Game already ended.');
    }

>     // ✅ Calcular todo o próximo estado no event handler

Reviewer: ```suggestion
    // ✅ Calcular todo o próximo state no manipulador de evento
```

  // ...
```

> Isso é muito mais eficiente. Além disso, se você implementar uma maneira de visualizar o histórico do jogo, agora poderá definir cada variável de estado para um movimento do passado sem acionar a cadeia de Effects que ajusta todos os outros valores. Se precisar reutilizar a lógica entre vários event handlers, você poderá [extrair uma função](#sharing-logic-between-event-handlers) e chamá-la a partir desses manipuladores.

Reviewer: ```suggestion
Isso é muito mais eficiente. Além disso, se você implementar uma maneira de visualizar o histórico do jogo, agora poderá definir cada variável de state para um movimento do passado sem acionar a cadeia de Effects que ajusta todos os outros valores. Se precisar reutilizar a lógica entre vários manipuladores de evento, você poderá [extrair uma função](#sharing-logic-between-event-handlers) e chamá-la a partir desses manipuladores.
```

> Lembre-se de que, dentro dos event handlers, [o estado se comporta como uma snapshot](/learn/state-as-a-snapshot). Por exemplo, mesmo depois de chamar `setRound(round + 1)`, a variável `round` refletirá o valor no momento em que o usuário clicou no botão. Se você precisar usar o próximo valor para cálculos, defina-o manualmente como `const nextRound = round + 1`.

Reviewer: ```suggestion
Lembre-se de que, dentro dos manipuladores de evento, [o estado se comporta como uma snapshot](/learn/state-as-a-snapshot). Por exemplo, mesmo depois de chamar `setRound(round + 1)`, a variável `round` refletirá o valor no momento em que o usuário clicou no botão. Se você precisar usar o próximo valor para cálculos, defina-o manualmente como `const nextRound = round + 1`.
```

> Em alguns casos, você *não pode* calcular o próximo estado diretamente no event handler. Por exemplo, imagine um formulário com vários dropdowns em que as opções do próximo dropdown dependem do valor selecionado do dropdown anterior. Nesse caso, uma cadeia de Effects é apropriada porque você está sincronizando com a rede.

Reviewer: ```suggestion
Em alguns casos, você *não pode* calcular o próximo state diretamente no manipulador de evento. Por exemplo, imagine um formulário com vários menus suspensos em que as opções do próximo menu dependem do valor selecionado do menu anterior. Nesse caso, uma cadeia de Effects é apropriada porque você está sincronizando com a rede.
```

> ### Inicialização do aplicativo {/*initializing-the-application*/}

Author: ```suggestion
### Inicialização da aplicação {/*initializing-the-application*/}
```

}
```

No entanto, você descobrirá rapidamente que ele [é executado duas vezes no desenvolvimento.](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development) Isso pode causar problemas - por exemplo, talvez ele invalide o token de autenticação porque a função não foi projetada para ser chamada duas vezes. Em geral, seus componentes devem ser resistentes à remontagem. Isso inclui seu componente `App` de nível superior.

> Embora talvez ele nunca seja remontado na prática em produção, seguir as mesmas restrições em todos os componentes facilita a movimentação e a reutilização do código. Se alguma lógica precisar ser executada *uma vez por carregamento de aplicativo* em vez de *uma vez por montagem de componente*, adicione uma variável no nível mais alto para registrar se ela já foi executada:

Author: ```suggestion
Embora talvez ele nunca seja remontado na prática em produção, seguir as mesmas restrições em todos os componentes facilita a movimentação e a reutilização do código. Se alguma lógica precisar ser executada *uma vez por carregamento da aplicação* em vez de *uma vez por montagem de componente*, adicione uma variável no nível mais alto para registrar se ela já foi executada:
```

  useEffect(() => {
    if (!didInit) {
      didInit = true;

>       // ✅ Só é executado uma vez por execução do aplicativo

Author: ```suggestion
      // ✅ Só é executado uma vez por execução do aplicativo
```
```suggestion
      // ✅ Só é executado uma vez por execução da aplicação
```

}
```

> Você também pode executá-lo durante a inicialização do módulo e antes de o aplicativo ser renderizado:

Author: ```suggestion
Você também pode executá-lo durante a inicialização do módulo e antes de a aplicação ser renderizada:
```

```js {1,5}
if (typeof window !== 'undefined') { // Verifica se estamos executando no navegador.

>    // ✅ Só é executado uma vez por execução do aplicativo

Author: ```suggestion
   // ✅ Só é executado uma vez por execução da aplicação
```

}
```

> O código no nível mais alto é executado uma vez quando o componente é importado, mesmo que ele não seja renderizado. Para evitar lentidão ou comportamento inesperado ao importar componentes arbitrários, não use esse padrão em excesso. Mantenha a lógica de inicialização de todo o aplicativo em módulos de componentes raiz como `App.js` ou no ponto de entrada do aplicativo.

Author: ```suggestion
O código no nível mais alto é executado uma vez quando o componente é importado, mesmo que ele não seja renderizado. Para evitar lentidão ou comportamento inesperado ao importar componentes arbitrários, não use esse padrão em excesso. Mantenha a lógica de inicialização de toda a aplicação em módulos de componentes raiz como `App.js` ou no ponto de entrada da aplicação.
```

### Notificar componentes pai sobre alterações de estado {/*notifying-parent-components-about-state-changes*/}

> Digamos que você esteja escrevendo um componente `Toggle` com um estado interno `isOn` que pode ser `true` ou `false`. Há algumas maneiras diferentes de alterná-lo (clicando ou arrastando). Você deseja notificar o componente pai sempre que o estado interno do `Toggle` for alterado, portanto, você expõe um evento `onChange` e o chama a partir de um Effect:

Reviewer: ```suggestion
Digamos que você esteja escrevendo um componente `Toggle` com um state interno `isOn` que pode ser `true` ou `false`. Há algumas maneiras diferentes de alterná-lo (clicando ou arrastando). Você deseja notificar o componente pai sempre que o estado interno do `Toggle` for alterado, portanto, você expõe um evento `onChange` e o chama a partir de um Effect:
```

```js {4-7}
function Toggle({ onChange }) {
  const [isOn, setIsOn] = useState(false);

>   // 🔴 Evitar: O handler onChange é executado tarde demais

Reviewer: ```suggestion
  // 🔴 Evitar: O manipulador onChange é executado tarde demais
```

}
```

> Como anteriormente, isso não é ideal. O `Toggle` atualiza seu estado primeiro, e o React atualiza a tela. Em seguida, o React executa o Effect, que chama a função `onChange` passada de um componente pai. Agora o componente pai atualizará seu próprio estado, iniciando outra passagem de renderização. Seria melhor fazer tudo em uma única passagem.

Reviewer: ```suggestion
Como anteriormente, isso não é ideal. O `Toggle` atualiza seu state primeiro, e o React atualiza a tela. Em seguida, o React executa o Effect, que chama a função `onChange` passada de um componente pai. Agora o componente pai atualizará seu próprio state, iniciando outra passagem de renderização. Seria melhor fazer tudo em uma única passagem.
```

> Exclua o Effect e, em vez disso, atualize o estado de *ambos* os componentes no mesmo event handler:

Reviewer: ```suggestion
Exclua o Effect e, em vez disso, atualize o state de *ambos* os componentes no mesmo manipulador de evento:
```

}
```

> Com essa abordagem, tanto o componente `Toggle` quanto seu componente pai atualizam seu estado durante o evento. O React [processa em lote atualizações](/learn/queueing-a-series-of-state-updates) de diferentes componentes juntos, de modo que haverá apenas uma passagem de renderização.

Reviewer: ```suggestion
Com essa abordagem, tanto o componente `Toggle` quanto seu componente pai atualizam seu state durante o evento. O React [processa em lote atualizações](/learn/queueing-a-series-of-state-updates) de diferentes componentes juntos, de modo que haverá apenas uma passagem de renderização.
```

> Você também pode remover completamente o estado e, em vez disso, receber `isOn` do componente pai:

Reviewer: ```suggestion
Você também pode remover completamente o state e, em vez disso, receber `isOn` do componente pai:
```

}
```

> ["Elevar o estado"](/learn/sharing-state-between-components) permite que o componente pai controle totalmente o `Toggle` alternando o estado do próprio componente pai. Isso significa que o componente pai terá que conter mais lógica, mas haverá menos estado geral com o qual se preocupar. Sempre que você tentar manter duas variáveis de estado diferentes sincronizadas, tente elevar o estado em vez disso!

Reviewer: ```suggestion
["Elevar o state"](/learn/sharing-state-between-components) permite que o componente pai controle totalmente o `Toggle` alternando o state do próprio componente pai. Isso significa que o componente pai terá que conter mais lógica, mas haverá menos state geral com o qual se preocupar. Sempre que você tentar manter duas variáveis de state diferentes sincronizadas, tente elevar o state em vez disso!
```

}
```

> No React, os dados fluem dos componentes pai para seus filhos. Quando você vê algo errado na tela, pode rastrear a origem da informação subindo a cadeia de componentes até encontrar o componente que passa a prop errada ou tem o estado errado. Quando os componentes filhos atualizam o estado de seus componentes pais em Effects, o fluxo de dados se torna muito difícil de rastrear. Como tanto o componente filho quanto o pai precisam dos mesmos dados, deixe o componente pai buscar esses dados e, em vez disso, *passá-los* para o filho:

Reviewer: ```suggestion
No React, os dados fluem dos componentes pai para seus filhos. Quando você vê algo errado na tela, pode rastrear a origem da informação subindo a cadeia de componentes até encontrar o componente que passa a prop errada ou tem o state errado. Quando os componentes filhos atualizam o state de seus componentes pais em Effects, o fluxo de dados se torna muito difícil de rastrear. Como tanto o componente filho quanto o pai precisam dos mesmos dados, deixe o componente pai buscar esses dados e, em vez disso, *passá-los* para o filho:
```

}
```

Isso é mais simples e mantém o fluxo de dados previsível: os dados fluem do pai para o filho.

### Inscrição em um armazenamento externo {/*subscribing-to-an-external-store*/}

> Às vezes, seus componentes podem precisar escutar alguns dados fora do estado do React. Esses dados podem ser de uma biblioteca de terceiros ou de uma API integrada do navegador. Como esses dados podem ser alterados sem o conhecimento do React, você precisa se inscrever manualmente em seus componentes. Isso geralmente é feito com um Effect, por exemplo:

Reviewer: ```suggestion
Às vezes, seus componentes podem precisar escutar alguns dados fora do state do React. Esses dados podem ser de uma biblioteca de terceiros ou de uma API integrada do navegador. Como esses dados podem ser alterados sem o conhecimento do React, você precisa se inscrever manualmente em seus componentes. Isso geralmente é feito com um Effect, por exemplo:
```

}
```

> Aqui, o componente se inscreve em um armazenamento de dados externo (nesse caso, a API `navigator.onLine` do navegador). Como essa API não existe no servidor (portanto, não pode ser usada para o HTML inicial), inicialmente o estado é definido como `true`. Sempre que o valor desse armazenamento de dados for alterado no navegador, o componente atualizará seu estado.

Reviewer: ```suggestion
Aqui, o componente se inscreve em um armazenamento de dados externo (nesse caso, a API `navigator.onLine` do navegador). Como essa API não existe no servidor (portanto, não pode ser usada para o HTML inicial), inicialmente o state é definido como `true`. Sempre que o valor desse armazenamento de dados for alterado no navegador, o componente atualizará seu estado.
```

}
```

> Essa abordagem é menos propensa a erros do que a sincronização manual de dados mutáveis para o estado do React com um Effect. Normalmente, você escreverá um Hook personalizado como o `useOnlineStatus()` acima para não precisar repetir esse código nos componentes individuais. [Leia mais sobre como assinar armazenamentos externos a partir de componentes React](/reference/react/useSyncExternalStore)

Reviewer: ```suggestion
Essa abordagem é menos propensa a erros do que a sincronização manual de dados mutáveis para o state do React com um Effect. Normalmente, você escreverá um Hook personalizado como o `useOnlineStatus()` acima para não precisar repetir esse código nos componentes individuais. [Leia mais sobre como assinar armazenamentos externos a partir de componentes React](/reference/react/useSyncExternalStore)
```

### Buscando dados {/*fetching-data*/}

> Muitos aplicativos usam o Effects para iniciar a busca de dados. É bastante comum escrever um Effect de busca de dados como este:

Author: ```suggestion
Muitas aplicações usam o Effects para iniciar a busca de dados. É bastante comum escrever um Effect de busca de dados como este:
```

}
```

> Você *não* precisa mover essa busca para um event handler.

Reviewer: ```suggestion
Você *não* precisa mover essa busca para um manipulador de evento.
```

}
```

Isso garante que, quando seu Effect buscar dados, todas as respostas, exceto a última solicitada, serão ignoradas.

Lidar com condições de corrida não é a única dificuldade na implementação da busca de dados. Talvez você também queira pensar em armazenar respostas em cache (para que o usuário possa clicar em Voltar e ver a tela anterior instantaneamente), como buscar dados no servidor (para que o HTML inicial renderizado pelo servidor contenha o conteúdo buscado em vez de um spinner) e como evitar cascatas de rede (para que um filho possa buscar dados sem esperar por todos os pais).

> **Esses problemas se aplicam a qualquer biblioteca de interface do usuário, não apenas ao React. Resolvê-los não é trivial, e é por isso que os [frameworks modernos](/learn/start-a-new-react-project#production-grade-react-frameworks) fornecem mecanismos internos de busca de dados mais eficientes do que a busca de dados no Effects.**

Reviewer: ```suggestion
**Esses problemas se aplicam a qualquer biblioteca de interface do usuário, não apenas ao React. Resolvê-los não é trivial, e é por isso que os [frameworks modernos](/learn/start-a-new-react-project#production-grade-react-frameworks) fornecem mecanismos internos de busca de dados mais eficientes do que a busca de dados nos Effects.**
```

> Se você não usa um framework (e não quer criar o sua próprio), mas gostaria de tornar a busca de dados do Effects mais ergonômica, considere extrair sua lógica de busca em um Hook personalizado, como neste exemplo:

Reviewer: ```suggestion
Se você não usa um framework (e não quer criar o seu próprio), mas gostaria de tornar a busca de dados dos Effects mais ergonômica, considere extrair sua lógica de busca em um Hook personalizado, como neste exemplo:
```

}
```

Provavelmente, você também desejará adicionar alguma lógica para tratamento de erros e para verificar se o conteúdo está sendo carregado. Você mesmo pode criar um Hook como esse ou usar uma das muitas soluções já disponíveis no ecossistema React. **Embora isso, por si só, não seja tão eficiente quanto usar o mecanismo de busca de dados integrado de um framework, mover a lógica de busca de dados para um Hook personalizado facilitará a adoção de uma estratégia eficiente de busca de dados posteriormente.**

> Em geral, sempre que tiver que recorrer à criação de Effects, fique atento para quando puder extrair uma parte da funcionalidade em um Hook personalizado com uma API mais declarativa e específica, como o `useData` acima. Quanto menos chamadas `useEffect` brutas você tiver em seus componentes, mais fácil será manter seu aplicativo.

Author: ```suggestion
Em geral, sempre que tiver que recorrer à criação de Effects, fique atento para quando puder extrair uma parte da funcionalidade em um Hook personalizado com uma API mais declarativa e específica, como o `useData` acima. Quanto menos chamadas `useEffect` brutas você tiver em seus componentes, mais fácil será manter sua aplicação.
```

<Recap>

- Se você puder calcular algo durante a renderização, não precisará de um Effect.
- Para armazenar em cache cálculos custosos, adicione `useMemo` em vez de `useEffect`.

> - Para redefinir o estado de uma árvore de componentes inteira, passe uma `key` diferente para ela.

Reviewer: ```suggestion
- Para redefinir o state de uma árvore de componentes inteira, passe uma `key` diferente para ela.
```

> - Para redefinir um determinado estado em resposta a uma alteração de prop, ajuste-o durante a renderização.

Reviewer: ```suggestion
- Para redefinir um determinado state em resposta a uma alteração de prop, ajuste-o durante a renderização.
```

- O código que é executado porque um componente foi *exibido* deve estar em Effects, o restante deve estar em eventos.

> - Se você precisar atualizar o estado de vários componentes, é melhor fazê-lo durante um único evento.

Reviewer: ```suggestion
- Se você precisar atualizar o state de vários componentes, é melhor fazê-lo durante um único evento.
```

> - Sempre que você tentar sincronizar variáveis de estado em diferentes componentes, considere elevar o estado.

Reviewer: ```suggestion
- Sempre que você tentar sincronizar variáveis de state em diferentes componentes, considere elevar o state.
```

- Você pode buscar dados com o Effects, mas precisa implementar a limpeza para evitar condições de corrida.

</Recap>

<Challenges>

#### Transformar dados sem Effects {/*transform-data-without-effects*/}

A `TodoList` abaixo exibe uma lista de todos. Quando a caixa de seleção "Show only active todos" está marcada, os *todos* concluídos não são exibidos na lista. Independentemente de quais *todos* estejam visíveis, o rodapé exibe a contagem de *todos* que ainda não foram concluídos.

> Simplifique esse componente removendo todo o estado e os Effects desnecessários.

Reviewer: ```suggestion
Simplifique esse componente removendo todo o state e os Effects desnecessários.
```

<Hint>

> Se você puder calcular algo durante a renderização, não precisará de estado ou de um Effect que o atualize.

Reviewer: ```suggestion
Se você puder calcular algo durante a renderização, não precisará de state ou de um Effect que o atualize.
```

</Hint>

<Solution>

> Há apenas duas partes essenciais de estado neste exemplo: a lista de `todos` e a variável de estado `showActive` que representa se a caixa de seleção está marcada. Todas as outras variáveis de estado são [redundantes](/learn/choosing-the-state-structure#avoid-redundant-state) e podem ser calculadas durante a renderização. Isso inclui o `footer`, que você pode mover diretamente para o JSX ao redor.

Reviewer: ```suggestion
Há apenas duas partes essenciais de estado neste exemplo: a lista de `todos` e a variável de state `showActive` que representa se a caixa de seleção está marcada. Todas as outras variáveis de state são [redundantes](/learn/choosing-the-state-structure#avoid-redundant-state) e podem ser calculadas durante a renderização. Isso inclui o `footer`, que você pode mover diretamente para a JSX circundante.
```

</Solution>

#### Armazenar em cache um cálculo sem Effects {/*cache-a-calculation-without-effects*/}

> Neste exemplo, a filtragem dos *todos* foi extraída em uma função separada chamada `getVisibleTodos()`. Essa função contém uma chamada `console.log()` dentro dela, o que o ajuda a perceber quando está sendo chamada. Alterne a opção "Show only active todos" e observe que isso faz com que a função `getVisibleTodos()` seja executada novamente. Isso é esperado porque os *todos* visíveis mudam quando você alterna quais devem ser exibidos.

Reviewer: ```suggestion
Neste exemplo, a filtragem dos *todos* foi extraída em uma função separada chamada `getVisibleTodos()`. Essa função contém uma chamada `console.log()` dentro dela, o que o ajuda a perceber quando está sendo chamada. Alterne a opção "Show only active todos" e perceba que isso faz com que a função `getVisibleTodos()` seja executada novamente. Isso é esperado porque os *todos* visíveis mudam quando você alterna quais devem ser exibidos.
```

<Solution>

> Remova a variável de estado e o Effect e, em vez disso, adicione uma chamada `useMemo` para armazenar em cache o resultado da chamada `getVisibleTodos()`:

Reviewer: ```suggestion
Remova a variável de state e o Effect e, em vez disso, adicione uma chamada `useMemo` para armazenar em cache o resultado da chamada `getVisibleTodos()`:
```

</Sandpack>

> Com essa alteração, a função `getVisibleTodos()` será chamada somente se `todos` ou `showActive` forem alterados. A digitação na entrada altera apenas a variável de estado `text`, portanto, não aciona uma chamada para `getVisibleTodos()`.

Reviewer: ```suggestion
Com essa alteração, a função `getVisibleTodos()` será chamada somente se `todos` ou `showActive` forem alterados. A digitação na entrada altera apenas a variável de state `text`, portanto, não aciona uma chamada para `getVisibleTodos()`.
```

> Há também outra solução que não precisa do `useMemo`. Como a variável de estado `text` não pode afetar a lista de todos, você pode extrair o formulário `NewTodo` em um componente separado e mover a variável de estado `text` para dentro dele:

Reviewer: ```suggestion
Há também outra solução que não precisa do `useMemo`. Como a variável de state `text` não pode afetar a lista de todos, você pode extrair o formulário `NewTodo` em um componente separado e mover a variável de estado `text` para dentro dele:
```

</Sandpack>

> Essa abordagem também satisfaz os requisitos. Quando você digita na entrada, apenas a variável de estado `text` é atualizada. Como a variável de estado `text` está no componente filho `NewTodo`, o componente pai `TodoList` não será renderizado novamente. É por isso que a função `getVisibleTodos()` não é chamada quando você digita. (Ele ainda seria chamado se o `TodoList` fosse renderizado novamente por outro motivo).

Reviewer: ```suggestion
Essa abordagem também satisfaz os requisitos. Quando você digita na entrada, apenas a variável de estado `text` é atualizada. Como a variável de state `text` está no componente filho `NewTodo`, o componente pai `TodoList` não será renderizado novamente. É por isso que a função `getVisibleTodos()` não é chamada quando você digita. (Ele ainda seria chamado se o `TodoList` fosse renderizado novamente por outro motivo).
```

</Solution>

> #### Redefinir estado sem Effects {/*reset-state-without-effects*/}

Reviewer: ```suggestion
#### Redefinir state sem Effects {/*reset-state-without-effects*/}
```

<Hint>

> Seria bom se houvesse uma maneira de dizer ao React que quando `savedContact.id` é diferente, o formulário `EditContact` é conceitualmente um _formulário de contato diferente_ e não deve preservar o estado. Você se lembra de alguma maneira?

Reviewer: ```suggestion
Seria bom se houvesse uma maneira de dizer ao React que quando `savedContact.id` é diferente, o formulário `EditContact` é conceitualmente um _formulário de contato diferente_ e não deve preservar o state. Você se lembra de alguma maneira?
```

</Hint>

<Solution>

> Divida o componente `EditContact` em dois. Mova todo o estado do formulário para o componente `EditForm` interno. Exporte o componente externo `EditContact` e faça com que ele passe `savedContact.id` como a `chave` para o componente interno `EditContact`. Como resultado, o componente interno `EditForm` redefine todo o estado do formulário e recria o DOM sempre que você seleciona um contato diferente.

Reviewer: ```suggestion
Divida o componente `EditContact` em dois. Mova todo o state do formulário para o componente `EditForm` interno. Exporte o componente externo `EditContact` e faça com que ele passe `savedContact.id` como a `chave` para o componente interno `EditContact`. Como resultado, o componente interno `EditForm` redefine todo o state do formulário e recria o DOM sempre que você seleciona um contato diferente.
```

</Solution>

#### Enviar um formulário sem Effects {/*submit-a-form-without-effects*/}

> Esse componente `Form` permite que você envie uma mensagem a um amigo. Quando você envia o formulário, a variável de estado `showForm` é definida como `false`. Isso aciona um Effect chamando `sendMessage(message)`, que envia a mensagem (você pode vê-la no console). Depois que a mensagem é enviada, você vê uma caixa de diálogo "Thank you" (Obrigado) com um botão "Open chat" (Abrir bate-papo) que permite que você volte ao formulário.

Reviewer: ```suggestion
Esse componente `Form` permite que você envie uma mensagem a um amigo. Quando você envia o formulário, a state de estado `showForm` é definida como `false`. Isso aciona um Effect chamando `sendMessage(message)`, que envia a mensagem (você pode vê-la no console). Depois que a mensagem é enviada, você vê uma caixa de diálogo "Thank you" (Obrigado) com um botão "Open chat" (Abrir bate-papo) que permite que você volte ao formulário.
```

Author: ```suggestion
Esse componente `Form` permite que você envie uma mensagem a um amigo. Quando você envia o formulário, a variável de state `showForm` é definida como `false`. Isso aciona um Effect chamando `sendMessage(message)`, que envia a mensagem (você pode vê-la no console). Depois que a mensagem é enviada, você vê uma caixa de diálogo "Thank you" (Obrigado) com um botão "Open chat" (Abrir bate-papo) que permite que você volte ao formulário.
```

> Os usuários do seu aplicativo estão enviando muitas mensagens. Para tornar o bate-papo um pouco mais difícil, você decidiu mostrar a caixa de diálogo "Thank you" *primeiro* em vez do formulário. Altere a variável de estado `showForm` para inicializar com `false` em vez de `true`. Assim que fizer essa alteração, o console mostrará que uma mensagem vazia foi enviada. Alguma coisa nessa lógica está errada!

Reviewer: ```suggestion
Os usuários do seu aplicativo estão enviando muitas mensagens. Para tornar o bate-papo um pouco mais difícil, você decidiu mostrar a caixa de diálogo "Thank you" *primeiro* em vez do formulário. Altere a variável de state `showForm` para inicializar com `false` em vez de `true`. Assim que fizer essa alteração, o console mostrará que uma mensagem vazia foi enviada. Alguma coisa nessa lógica está errada!
```

Author: ```suggestion
Os usuários da sua aplicação estão enviando muitas mensagens. Para tornar o bate-papo um pouco mais difícil, você decidiu mostrar a caixa de diálogo "Thank you" *primeiro* em vez do formulário. Altere a variável de state `showForm` para inicializar com `false` em vez de `true`. Assim que fizer essa alteração, o console mostrará que uma mensagem vazia foi enviada. Alguma coisa nessa lógica está errada!
```

<Solution>

> A variável de estado `showForm` determina se deve ser exibido o formulário ou a caixa de diálogo "Thank you". No entanto, você não está enviando a mensagem porque a caixa de diálogo "Thank you" foi _exibida_. Você deseja enviar a mensagem porque o usuário enviou o formulário. Exclua o Effect enganoso e mova a chamada `sendMessage` para dentro do event handler `handleSubmit`:

Reviewer: ```suggestion
A variável de state `showForm` determina se deve ser exibido o formulário ou a caixa de diálogo "Thank you". No entanto, você não está enviando a mensagem porque a caixa de diálogo "Thank you" foi _exibida_. Você deseja enviar a mensagem porque o usuário enviou o formulário. Exclua o Effect enganoso e mova a chamada `sendMessage` para dentro do manipulador de evento `handleSubmit`:
```

</Sandpack>

> Observe como, nessa versão, apenas _submeter o formulário_ (que é um evento) faz com que a mensagem seja enviada. Isso funciona igualmente bem, independentemente de `showForm` ser inicialmente definido como `true` ou `false`. (Defina-o como `false` e não observe nenhuma mensagem extra no console).

Reviewer: ```suggestion
Perceba como, nessa versão, apenas _submeter o formulário_ (que é um evento) faz com que a mensagem seja enviada. Isso funciona igualmente bem, independentemente de `showForm` ser inicialmente definido como `true` ou `false`. (Defina-o como `false` e não observe nenhuma mensagem extra no console).
```
